### PR TITLE
EventSource linker warnings fix

### PR DESCRIPTION
**Pull request by LakshanF**
First set of changes for EventSource trim warnings. This is a WIP that  fixes EventSource.cs warnings with other fixes to come

### CODE REVIEW

File: src/tests/Loader/binding/tracing/BinderTracingTest.ResolutionFlow.cs [hunk 1]
```diff
@@ -442,6 +442,7 @@ public static BindOperation AssemblyLoadContextResolvingEvent_CustomALC_Exceptio
         public static BindOperation AssemblyLoadContextResolvingEvent_DefaultALC_Exception()
         {
             var assemblyName = new AssemblyName(SubdirectoryAssemblyName);
+            System.Diagnostics.Debugger.Break();
```
Comment (AaronRobinsonMSFT) [inline src/tests/Loader/binding/tracing/BinderTracingTest.ResolutionFlow.cs:445]: If a debugger isn't attached I believe this will trigger an exception. Instead I would branch only if [`Debugger.IsAttached`](https://docs.microsoft.com/dotnet/api/system.diagnostics.debugger.isattached) is `true`.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 2]
```diff
@@ -2647,7 +2646,7 @@ private void EnsureDescriptorsInitialized()
             {
                 // get the metadata via reflection.
                 Debug.Assert(m_rawManifest == null);
-                m_rawManifest = CreateManifestAndDescriptors(this.GetType(), Name, this);
+                m_rawManifest = CreateManifestAndDescriptors(typeof(EventSource), Name, this);
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2649]: This needs to walk the members on types deriving from EventSource - I assume this is why the tests are failing.

This line needs to be kept as-is and we need something like https://github.com/mono/linker/issues/1196 in the linker.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 3]
```diff
@@ -334,8 +334,7 @@ public static string GetName(Type eventSourceType)
         }
 
 #if !ES_BUILD_STANDALONE
-        private const DynamicallyAccessedMemberTypes ManifestMemberTypes = DynamicallyAccessedMemberTypes.PublicNestedTypes
-            | DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods;
+        private const DynamicallyAccessedMemberTypes ManifestMemberTypes = DynamicallyAccessedMemberTypes.All;
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:337]: We wouldn't need to bump this to "keep the entire type" if we did https://github.com/mono/linker/pull/1719#issuecomment-753444772

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs (hunk 2 above)
Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs]: Thanks for catching that! I reverted this for the time being since the issue you point is still open

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs (hunk 3 above)
Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:337]: I took the suppression attribute out since that is how I understood Vitek's fix to be. Let me know if the fix will also allow DynamicallyAccessedMemberTypes.All annotation to be replaced with an appropriate subset (my test didn't show that to be the case)

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 4]
```diff
@@ -5372,6 +5379,9 @@ public void ManifestError(string msg, bool runtimeCritical = false)
                 throw new ArgumentException(msg);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+        Justification = "Calls to GetFields on types in mapsTab is safe " +
+        "since all its members are preserved")]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5384]:
```suggestion
        Justification = "Calls to GetFields on types in mapsTab are safe since all their members are preserved")]
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 5]
```diff
@@ -2879,7 +2878,10 @@ private static bool AttributeTypeNamesMatch(Type attributeType, Type reflectedAt
         // Use reflection to look at the attributes of a class, and generate a manifest for it (as UTF8) and
         // return the UTF8 bytes.  It also sets up the code:EventData structures needed to dispatch events
         // at run time.  'source' is the event source to place the descriptors.  If it is null,
-        // then the descriptors are not creaed, and just the manifest is generated.
+        // then the descriptors are not created, and just the manifest is generated.
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:UnrecognizedReflectionPattern",
+        Justification = "All eventSourceType's members in its entirety are preserved  " +
+        "with the parameter tagging")]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2884]:
```suggestion
            Justification = "All eventSourceType's members in their entirety are preserved with the parameter tagging")]
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 6]
```diff
@@ -5779,7 +5789,14 @@ private void AppendKeywords(StringBuilder sb, ulong keywords, string eventName)
             }
         }
 
-        private string GetTypeName(Type type)
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:UnrecognizedReflectionPattern",
+        Justification = "The recursive call GetType is safe since the type is not an enum in " +
+        "the 2nd call")]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5794]:
```suggestion
            Justification = "The recursive GetType call is safe since the type is not an enum in the 2nd call")]
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 7]
```diff
@@ -5372,6 +5379,9 @@ public void ManifestError(string msg, bool runtimeCritical = false)
                 throw new ArgumentException(msg);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5382]: Can we update the ILLink.Suppressions.xml file as we make these changes? That way the file stays up to date, and we can see what other ILLink warnings are left in this area.

https://github.com/dotnet/runtime/blob/3e1e1c15a73301e34c96b086764da7b7802514f9/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Suppressions.Shared.xml#L136-L141

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 8]
```diff
@@ -2879,7 +2878,10 @@ private static bool AttributeTypeNamesMatch(Type attributeType, Type reflectedAt
         // Use reflection to look at the attributes of a class, and generate a manifest for it (as UTF8) and
         // return the UTF8 bytes.  It also sets up the code:EventData structures needed to dispatch events
         // at run time.  'source' is the event source to place the descriptors.  If it is null,
-        // then the descriptors are not creaed, and just the manifest is generated.
+        // then the descriptors are not created, and just the manifest is generated.
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:UnrecognizedReflectionPattern",
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2882]: This is kind of a large method to be suppressing `IL2072` for the whole thing. Would it be better if we refactored out just the one place that needs to be suppressed in a separate method?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 9]
```diff
@@ -5246,7 +5248,12 @@ public void StartEvent(string eventName, EventAttribute eventAttribute)
 #endif
         }
 
-        public void AddEventParameter(Type type, string name)
+        public void AddEventParameter(
+#if !ES_BUILD_STANDALONE
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5253]: I'm not sure I understand why this is necessary. This method doesn't appear to use reflection on `type` and I don't see any warnings in the suppressions file.

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5253]: Ah, I see it calls `GetTypeName` which needs `DynamicallyAccessedMemberTypes.Fields`. We should probably be matching the MemberTypes, if possible, so more things aren't preserved than are necessary.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 10]
```diff
@@ -5779,7 +5789,14 @@ private void AppendKeywords(StringBuilder sb, ulong keywords, string eventName)
             }
         }
 
-        private string GetTypeName(Type type)
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:UnrecognizedReflectionPattern",
+        Justification = "The recursive call GetType is safe since the type is not an enum in " +
+        "the 2nd call")]
+        private string GetTypeName(
+#if !ES_BUILD_STANDALONE
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields | DynamicallyAccessedMemberTypes.NonPublicFields)]
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5797]: Can we just suppress the warning for `type.GetFields` here? I don't think the ILLinker can ever remove the `value__` field from an Enum.

Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5797]: Can we just make this call `Type.GetEnumUnderlyingType`?

EventSource is basically inlining what the default implementation of this is:

https://github.com/dotnet/runtime/blob/867ba801d1a609f4cbb6e5847a3ec030adb2d8f4/src/libraries/System.Private.CoreLib/src/System/Type.cs#L485-L495

And it skips any possible optimizations the RuntimeType might have.

https://github.com/dotnet/runtime/blob/867ba801d1a609f4cbb6e5847a3ec030adb2d8f4/src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs#L145-L151

It's odd to have odd code that we then need to create suppressions for.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs (hunk 9 above)
Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs]: I don’t think this change is needed.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 11]
```diff
@@ -3089,13 +3088,18 @@ private static bool AttributeTypeNamesMatch(Type attributeType, Type reflectedAt
                             }
                         }
 
+                        // Scoping the call to AddEventParameter to a local function to limit the linker suppression
+                        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2067:UnrecognizedReflectionPattern",
+                        Justification = "All eventSourceType's members in their entirety are preserved with the parameter tagging")]
+                        void CallAddEventParameter(Type type, string name) => manifest.AddEventParameter(type, name);
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:3094]: Do we still need this now that `AddEventParameter` is no longer attributed?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 12]
```diff
@@ -5372,6 +5376,8 @@ public void ManifestError(string msg, bool runtimeCritical = false)
                 throw new ArgumentException(msg);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+        Justification = "Calls to GetFields on types in mapsTab are safe since all their members are preserved")]
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: I would reword/clarify this justification. The call to GetFields is getting all of the enum values of an enum type.

Will the ILLinker ever trim unused enum values from an enum? My guess is "no", but it would be good to verify that. If it is "no", then we should say that in the justification.

Similar to the above comment, I would move the suppress message to just the one place that needs to be suppressed.

And then similar to @MichalStrehovsky's suggestion, we could change the use of Reflection with Enum.GetValues and Enum.GetNames.

Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: > And then similar to @MichalStrehovsky's suggestion, we could change the use of Reflection with Enum.GetValues and Enum.GetNames.

`Enum.GetValues` creates an array of the enum type. I'm not sure if this works in ReflectionOnly context. EventSource tries to work in ReflectionOnly context for scenarios I don't know much about. I don't know if it's needed here, but I thought I would call that out in case one of the other reviewers knows if that's acceptable. It's definitely preferred because it doesn't need annotations.

Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: It seems like `Enum.GetValues` can be used since the types are explicitly checked to be an `enum` before adding to the collection, `mapsTab` and that would not require annotations or suppressions.

Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: > It seems like `Enum.GetValues` can be used since the types are explicitly checked to be an `enum` before adding to the collection, `mapsTab` and that would not require annotations or suppressions. 

To make it clear what I'm hinting at:

```csharp
using System;
using System.Reflection;

Type type = new TypeDelegator(typeof(MyEnum));
Console.WriteLine(type.IsEnum);
Console.WriteLine(new string('-', 80));
foreach (var f in type.GetFields())
    Console.WriteLine(f);
Console.WriteLine(new string('-', 80));
foreach (var v in Enum.GetValues(type))
    Console.WriteLine(v);


enum MyEnum { One, Two }
```

This will print:

```
True
--------------------------------------------------------------------------------
Int32 value__
MyEnum One
MyEnum Two
--------------------------------------------------------------------------------
Unhandled exception. System.NotImplementedException: The method or operation is not implemented.
   at System.Type.GetEnumValues()
   at <Program>$.<Main>$(String[] args) in C:\Users\michals\source\repos\NetCoreConsoleApp1\Program.cs:line 14
```

Does this codepath need to operate in such environment?

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: > Does this codepath need to operate in such environment?

I don't _think_ so.... but we should verify. Looking at how this code uses `FieldInfo`:

https://github.com/dotnet/runtime/blob/cb0e8b474a8f26b43566424d98f3014a0b874a71/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L5466-L5468

```C#
using System;
using System.Reflection;

Type type = new TypeDelegator(typeof(MyEnum));
Console.WriteLine(type.IsEnum);
Console.WriteLine(new string('-', 80));
foreach (var f in type.GetFields())
    Console.WriteLine(f.GetRawConstantValue());

enum MyEnum { One, Two }
```

 will print:

```
True
--------------------------------------------------------------------------------
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at System.Reflection.RtFieldInfo.GetRawConstantValue()
   at Submission#7.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)
   at System.Reflection.RtFieldInfo.GetRawConstantValue()
   at Submission#7.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)
```

Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: I see, cannot say definitively if this can be ruled out. Sticking to the local function suppression seems safest. Thanks!

Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5380]: Unfortunately, the current code has `BindingFlags` and that will work.

```
using System;
using System.Reflection;

Type type = new TypeDelegator(typeof(MyEnum));
Console.WriteLine(type.IsEnum);
Console.WriteLine(new string('-', 80));
foreach (var f in type.GetFields(BindingFlags.DeclaredOnly | BindingFlags.Public | BindingFlags.Static))
    Console.WriteLine(f.GetRawConstantValue());

enum MyEnum { One, Two }
```

will work:
```
True
--------------------------------------------------------------------------------
0
1
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 13]
```diff
@@ -2939,6 +2938,11 @@ private static bool AttributeTypeNamesMatch(Type attributeType, Type reflectedAt
                     }
                 }
 
+                // Scoping the call to GetFields to a local function to limit the linker suppression
+                [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+                Justification = "Nested type members are safe from linker with the parent annotation")]
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2943]: (nit) The `Justification` line should be indented one level to the right.

(nit) "safe from linker" sounds a bit like the linker is dangerous. How about "Nested type members will be preserved by the ILLinker with the parent annotation" ?

Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs]: Or just skip the "linker" part altogether. "Nested type members will be preserved with the parent annotation"

.NET is trying to overload what the term [linker](https://en.wikipedia.org/wiki/Linker_(computing)) means in computing. That is never a good idea. IL Linker is not doing any of what anyone outside .NET thinks of when they hear "linker". We should stay away from terms linking/linker as much as possible and use the term "trimming".

We are introducing unnecessary confusion into .NET with naming the tool "linker". It's hard to go back on the naming there, but we can make conscious effort to not have the name leak into other parts of .NET. The user-visible parts of project settings already do a pretty good job of avoiding the unfortunate name.

**Review by eerhardt — APPROVED** [review]
LGTM

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 14]
```diff
@@ -2936,6 +2935,13 @@ private static bool AttributeTypeNamesMatch(Type attributeType, Type reflectedAt
                     }
                 }
 
+                // Scoping the call to GetFields to a local function to limit the linker suppression
+#if !ES_BUILD_STANDALONE
+                [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+                    Justification = "Nested type members will be preserved with the parent annotation")]
+#endif
```
Comment (vitek-karas) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2942]: With the new linker this should not be necessary (both the suppression and local function)

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 15]
```diff
@@ -5451,6 +5457,18 @@ private string CreateManifestString()
             }
 
             // Write out the maps
+
+            // Scoping the call to enum GetFields to a local function to limit the linker suppression
+#if !ES_BUILD_STANDALONE
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "Linker does not trim enums")]
```
Comment (vitek-karas) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:5464]: Nit: Try to avoid "linker" in the codebase - "trimmer" is a good replacement.

**Review by vitek-karas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs (hunk 14 above)
Comment (LakshanF) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2942]: Is the new linker picked in the repo builds? I added cause the checks failed in the PR build (before you checked in the code though)

Comment (vitek-karas) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2942]: Yes - it's auto updated - I think this is the commit which brought it into runtime... but I could be wrong: https://github.com/dotnet/runtime/commit/2f03850bade9930757281f8840b5c6745f3f9a3a
